The fourth milestone of the Learner Dashboard project.

arunabh98 opened the pull request:
The feedback feature is implemented.

arunabh98 committed:
Fetch feedback summary to the dashboard.

arunabh98 committed:
fix some bugs.

arunabh98 committed:
Fetch thread summaries.

arunabh98 committed:
remove unused functions.

arunabh98 committed:
add tests.

arunabh98 committed:
Fix lint errors.

arunabh98 committed:
fix more lint errors.

arunabh98 committed:
Add test for get all threads.

arunabh98 committed:
Complete summary.

arunabh98 committed:
Add handler test.

arunabh98 committed:
Remove unnecessary lines.

arunabh98 committed:
Add overall test and address review comments.

arunabh98 committed:
Add load test.

arunabh98 committed:
message count added to thread model.

arunabh98 committed:
Estimate message count by id.

arunabh98 committed:
Merge branch 'learner-dashboard-milestone-third' into develop

arunabh98 committed:
Fetch thread summaries.

arunabh98 committed:
Display feedback threads on the dashboard.

arunabh98 committed:
address review comments.

arunabh98 committed:
minor nits.

arunabh98 committed:
fix lint errors.

arunabh98 committed:
fix lint errors.

arunabh98 committed:
Complete work on feedback threads.

arunabh98 committed:
start working on suggestions.

arunabh98 committed:
Address review comments 2.

arunabh98 committed:
Fix minor errors.

arunabh98 committed:
address review comments 3.

arunabh98 committed:
Added back index.

arunabh98 committed:
Add tests.

arunabh98 committed:
address review comments.

arunabh98 committed:
fix typos.

arunabh98 committed:
Fetch current state content.

arunabh98 committed:
Add current content to suggestion modal.

arunabh98 committed:
Fix merge conflicts.

arunabh98 committed:
Add e2e tests and some more features.

arunabh98 committed:
Show unread threads and wrote e2e test.

arunabh98 committed:
Fix lint errors.

arunabh98 committed:
Fix more lint errors.

arunabh98 committed:
Fix code style.

seanlip commented on core/controllers/learner_dashboard.py:
```diff
@@ -73,6 +74,16 @@ def get(self):
+                feedback_services.get_thread_summaries(
+                    self.user_id, full_thread_ids))
+        else:
+            thread_summaries = []
```
For parity why not drop number_of_unread_threads = 0 from line 79 and do `thread_summaries, number_of_unread_threads = [], 0`?

arunabh98 replied:
Done!

seanlip commented on core/controllers/learner_dashboard.py:
```diff
@@ -91,6 +102,8 @@ def get(self):
 
             subscription_list.append(subscription_summary)
 
+        user_settings = user_services.get_user_settings(self.user_id)
```
Just wondering: I thought we already load the user's profile picture and username on every page already? Not sure I understand why we need to get it again here.

arunabh98 replied:
Done! Didn't know that.

seanlip commented on core/controllers/learner_dashboard.py:
```diff
@@ -122,12 +139,32 @@ def get(self, exploration_id, thread_id):
+            exploration = exp_services.get_exploration_by_id(exploration_id)
+            current_content_html = (
+                exploration.states[
+                    suggestion.state_name].content.to_dict()['html'])
```
Seems a bit indirect; couldn't we do something like content.get_html() or content.html or something like that? Not sure why converting to a dict is necessary.

arunabh98 replied:
Done!

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -94,7 +94,7 @@ def test_message_count(self):
         self._run_one_off_job()
 
-        thread_summaries = feedback_services.get_thread_summaries(
+        thread_summaries, _ = feedback_services.get_thread_summaries(
```
Well, you might as well test the second arg... (probably in feedback_services_test, though, I guess).

In general please add meaningful tests for all new functionality. I _think_ I've commented on this in several PRs now, so in the future, this should go without saying. You want your code to not break in the future!

arunabh98 replied:
Done! Sure!

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -508,7 +514,13 @@ def get_thread_summaries(user_id, full_thread_ids):
-            author_second_last_message = last_two_messages[index][1].author_id
+            author_second_last_message = user_services.get_username(
+                last_two_messages[index][1].author_id)
+            if not last_message_read or not second_last_message_read:
```
This slightly confuses me. In what case would the last message have been read, but the second last message be not read? If that case is impossible then this part could presumably be pulled out of both the if and else clauses into the main execution flow.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/collection_player/collection_player.html:
```diff
@@ -153,7 +153,8 @@ <h2 ng-if="collectionPlaythrough.hasFinishedCollection()" class="oppia-page-head
                version="1.1"
-               ng-attr-style="position: absolute; left: <[pathIconParameters[$index].left]>; top: <[pathIconParameters[$index].top]>;">
+               ng-attr-style="position: absolute; left: <[pathIconParameters[$index].left]>; top: <[pathIconParameters[$index].top]>;"
+               class="protractor-test-collection-exploration">
```
I actually have no idea what protractor-test-collection-exploration means. Perhaps protractor-test-collection-node?

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -49,30 +55,51 @@ oppia.constant('HUMAN_READABLE_SUBSCRIPTION_SORT_BY_KEYS', {
   IMPACT: 'Impact'
 });
 
+oppia.constant('HUMAN_READABLE_FEEDBACK_THREADS_SORT_BY_KEYS', {
```
Instead of separating this from the above, perhaps it might be better to have a constant that looks something like this:

```
{
  LAST_UPDATED: {
    key: 'last_updated',
    i18nId: 'I18N_...'
  },
  ...
}
```

so that the two things automatically stay in sync. What do you think? Also I would specifically refer to i18n IDs here rather than "human readable string" because 'I18N_DASHBOARD_...' is not a human-readable string.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
 
+    $scope.sortFeedbackFunction = function(entity) {
+      return entity[$scope.currentFeedbackThreadsSortType];
+    }
```
Missing semicolon.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
       return value;
     };
 
+    $scope.sortFeedbackFunction = function(entity) {
```
I have no idea what `entity` means here, and I also don't know what "sortFeedbackFunction" is supposed to do from the name. Could you use names that are more illustrative of what these actually are?

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
+
+    $scope.onClickThread = function(
+      explorationId, threadId, explorationTitle) {
+      var threadDataUrl = (
```
Use UrlInterpolationService.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
+      explorationId, threadId, explorationTitle) {
+      var threadDataUrl = (
+        '/learnerdashboardthreadhandler/' + explorationId + '/' + threadId);
+      $scope.threadTitle = explorationTitle;
```
It seems weird to me that the threadTitle is the explorationTitle. If you want to display the latter, why not just call it explorationTitle?

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
+      for (var index = 0; index < $scope.threadSummaries.length; index++) {
+        if ($scope.threadSummaries[index].exploration_id === explorationId &&
+            $scope.threadSummaries[index].thread_id === threadId) {
+          $scope.threadIndex = index;
```
Where's $scope.threadIndex defined? Where is it set back to null?

Also, if you don't use it in the HTML then use 'var' instead to keep it private to the controller.

This controller is getting long, btw. Try pulling out coherent subparts of it into a directive, if you can.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
+        if ($scope.threadSummaries[index].exploration_id === explorationId &&
+            $scope.threadSummaries[index].thread_id === threadId) {
+          $scope.threadIndex = index;
+          var threadSummary = $scope.threadSummaries[index]
```
Missing semicolon.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
+            $scope.threadSummaries[index].thread_id === threadId) {
+          $scope.threadIndex = index;
+          var threadSummary = $scope.threadSummaries[index]
+          if (typeof threadSummary.second_last_message_read !== 'undefined') {
```
Why not just `if (threadSummary.second_last_message_read)`?

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
+          if (typeof threadSummary.second_last_message_read !== 'undefined') {
+            $scope.threadSummaries[index].second_last_message_read = true;
+          }
+          if (threadSummary.last_message_read === false) {
```
if (!threadSummary.last_message_read)

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
+      });
+    };
+
+    $scope.backFromThread = function() {
```
Function names should start with verbs.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
+    };
+
+    $scope.addNewMessage = function(explorationId, threadId, newMessage) {
+      var url = '/threadhandler/' + explorationId + '/' + threadId;
```
Use UrlInterpolationService.

arunabh98 replied:
I had earlier used this but then I get an error message like this:
![image](https://user-images.githubusercontent.com/13691317/28447804-e3da39ca-6df0-11e7-8c5e-3a65834b5365.png).
The thread id contains characters that are not allowed. What should we do about this?

seanlip replied:
I think it should be fine in this case to extend the range of characters used by UrlInterpolationService. (I assume we don't need to url-encode stuff in this case?)

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
+      };
+      $scope.messageSendingInProgress = true;
+      $http.post(url, payload).success(function() {
+        var messageSummary = {};
```
You ought to have a domain object for MessageSummary that you use in the frontend, rather than passing the backend stuff around. See the exploration editor for how this is done with e.g. State, Interaction, etc.; @tjiang11 can probably give you advice on this if you have questions.

Ditto for ThreadSummary etc.

arunabh98 replied:
Could you please explain this a bit further? Here I've just temporarily updated the summaries so that if the user clicks return to feedback threads he/she is able to see updated fields. If the user reloads the learner dashboard the summaries come naturally as we have sent a post request above having a payload that only contains the new message.

seanlip replied:
Sure. Take a look at e.g. [ExplorationObjectFactory](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/domain/exploration/ExplorationObjectFactory.js), for instance, and see how it's used in the codebase. Basically, when a response comes back from the backend, that dict is converted into a domain object that then subsequently gets used in the frontend logic. It has helper methods for common operations and is basically a plain JavaScript object. In general, the frontend logic should be working with domain objects and not the raw dicts from the backend.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -19,7 +19,8 @@
 oppia.constant('LEARNER_DASHBOARD_SECTIONS', {
```
LEARNER_DASHBOARD_SECTION_I18N_IDS

(Ditto for other constants below.)

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/learner_dashboard.html:
```diff
@@ -440,6 +463,119 @@ <h1 class="stat-value"><[subscriptionsList.length]></h1>
+                    <strong><[thread.exploration_title]> -</strong>
+                  </span>
+                  <span ng-if="thread.last_message_text"><[thread.last_message_text | truncate:50]></span>
+                  <span ng-if="!thread.last_message_text"><i>Suggestion</i></span>
```
Does "Suggestion" need to be internationalized? Are there other strings that need to be?

arunabh98 replied:
Internationalized everything!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/learner_dashboard.html:
```diff
@@ -440,6 +463,119 @@ <h1 class="stat-value"><[subscriptionsList.length]></h1>
+                  <button ng-click="backFromThread()"
+                          class="btn btn-default"
+                          style="padding: 0 3px;">
+                    <i class="material-icons oppia-vcenter" title="Return to list of messages"></i>
```
Not sure why there's an unreadable character here. Shouldn't you use a code like `&#x<[something]>;`, following the [Material Icons site](https://material.io/icons/)?

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/learner_dashboard.html:
```diff
@@ -440,6 +463,119 @@ <h1 class="stat-value"><[subscriptionsList.length]></h1>
+
+                    <div class="col-sm-11">
+                      <div>
+                        <textarea class="form-control feedback-thread-reply-box" placeholder="Reply" ng-model="newMessage"></textarea>
```
Should "reply" be i18n'd? Brings up an interesting question -- I guess we expect messages that people write to be in English? Does it depend on the language of the exploration?

arunabh98 replied:
Internationalized.

seanlip replied:
I don't think you answered my questions...

seanlip commented on core/templates/dev/head/pages/learner_dashboard/learner_dashboard.html:
```diff
@@ -469,6 +605,40 @@ <h1 class="stat-value"><[subscriptionsList.length]></h1>
     </div>
   </script>
+
+  <script type="text/ng-template" id="modals/editorViewSuggestion">
```
Is this the same as the suggestion modal in the editor? If not, why call it editorViewSuggestion (since this is a learner-facing dashboard)?

arunabh98 replied:
Done!

seanlip replied:
A note for the future: for comments like this (that contain questions), please answer the questions, or explain what you did. This makes the review go faster. Thanks!

seanlip commented on core/tests/protractor/learnerDashboard.js:
```diff
@@ -168,7 +168,33 @@ describe('Learner dashboard functionality', function() {
-  });  
+  });
+
+  it('display learner feedback threads', function() {
```
it('displays ...'

(make it read like a sentence)

arunabh98 replied:
Done!

seanlip reviewed (changes requested):
Thanks, @Arunabh98! I looked only at the code; I defer to your mentors @jaredsilver and @rachelwchen for UI/UX review.

arunabh98 committed:
Partly address review comments.:

jaredsilver reviewed (changes requested):
Thanks, Arunabh!

One initial edge case that pops up is what happens if a user starts an exploration and leaves feedback on the first card but doesn't switch cards. This results in the "It looks like you haven't tried any of our explorations yet." message on the dashboard, meaning the user cannot access feedback threads. This probably isn't important to fix, but it's worth considering.

One major concern is how we're handling changing the status of a message. The creator feedback panel shows, for example, "Status changed to 'Ignored'," but the learner feedback panel does not show this message. The learner feedback panel instead shows a blank message (or the contents of the textbox if there was a message accompanying the status change from the creator). I guess we could just show the status change text in the learner feedback panel as well?

The other concern with the statuses is that we _are_ showing them in the index of conversations on the learner feedback panel, and some of the statuses are negative (e.g. 'Ignored'). This might create a negative experience for the learner, so perhaps we should have these display just as "open" or "closed" (with no additional information about why it was closed).

A small nitpick: if I'm in a specific thread, I'd expect to be able to click on "Feedback Updates" in the sidebar to be returned to the index of threads.

When a user sends a message, the text contents of that message appear again in the textbox. This should clear after a message is sent.

Can we disallow horizontal resizing of the textbox? The container doesn't grow with it, so the user might accidentally make it look like this:

<img width="914" alt="screen shot 2017-07-20 at 5 09 24 pm" src="https://user-images.githubusercontent.com/1312199/28441013-48531698-6d6e-11e7-8fd6-4661ac8aae0c.png">

Can we eliminate the margin-left of the table with `ng-if="!feedbackThreadActive"` for small viewports? It's causing the feedback container to move off the edge on small screens.

I'm wondering now if we may actually be better not including the username column here. Presumably, all communication is occurring just between the learner and any of the exploration creators, and these are identified in the thread itself. So long as the new message indicator is displayed, displaying the usernames might be unnecessary cognitive overhead.

I'm not certain, but I think that the sorting arrow might be pointing in the opposite direction with respect to time and alphabetization.

Thanks a bunch!

arunabh98 committed:
Fullt addres review comments.

arunabh98 committed:
Fix lint errors.

arunabh98 committed:
Fix more lint errors.

arunabh98 committed:
Fix fronend test error.

arunabh98 committed:
fix merge conflicts.

arunabh98 committed:
Partly address review comments 2.

arunabh98 committed:
Fully address review comments 2.

arunabh98 committed:
Fix backend tests.

arunabh98 commented:
Hi @jaredsilver, Thanks for your review!
I have handled the status change issue in the feedback threads. A message now appears which shows the user the status has been changed. I initially implemented the Closed for 'ignored' and 'not actionable' but on further thought, the creator will normally also provide an explanation along with these statuses which can provide valuable context and more information to the learner as to exactly what was missing in his/her feedback and why it was 'not actionable' or 'ignored'.

Implemented the clicking on feedback thread suggestion.

Disabled horizontal resizing of the textbox.

Eliminated left margin for small view ports.

Fixed the sorting arrow.

And I actually think user column looks nice and it's often helpful to know who sent the last message and whether we have read the messages or not. This is very similar to Gmail thread behavior.

Also regarding the first point, since it's not that important an issue and probably not that odd too, I'll tackle this later after I've completed the more important parts.

seanlip commented:
> I have handled the status change issue in the feedback threads. A message now appears which shows the user the status has been changed. I initially implemented the Closed for 'ignored' and 'not actionable' but on further thought, the creator will normally also provide an explanation along with these statuses which can provide valuable context and more information to the learner as to exactly what was missing in his/her feedback and why it was 'not actionable' or 'ignored'.

Just a note on this -- if this is our operating assumption then I think we need to add a note in the editor feedback tab that prohibits changing the status to one of these and not leaving a message (at least in the case where the feedback was submitted by a named user and not anonymously). A warning could appear next to the submit button there that says something like "please specify a reason" or similar. Otherwise the experience of having a piece of feedback ignored without further comment could be quite negative.

seanlip commented on assets/i18n/qqq.json:
```diff
@@ -216,11 +217,25 @@
 	"I18N_LEARNER_DASHBOARD_INTRO_MESSAGE_PART_ONE": "First part of the introductory message that appears on the learner dashboard when there is no activity.",
 	"I18N_LEARNER_DASHBOARD_INTRO_MESSAGE_PART_TWO": "Second part of the introductory message that appears on the learner dashboard when there is no activity.",
-	"I18N_LEARNER_DASHBOARD_SUBSCRIPTIONS_SECTION": "Text for the subscriptions section in the learner dashboard.",
+	"I18N_LEARNER_DASHBOARD_REMOVE_ACTIVITY_MODAL_BODY": "Text that appears in the body of the remove modal in the learner dashboard. It asks the user if he/she is sure to remove the activity from the learner dashboard.",
```
I think you should describe what entityTitle, sectionName are here. Otherwise the translators won't have any idea (they only see these files, not the site).

Also, looking at the translation files, it looks like sectionName is a misnomer, since it's being passed through the translation filter? Should it be sectionNameI18nId or something like that?

arunabh98 replied:
Done!

seanlip commented on core/controllers/learner_dashboard.py:
```diff
@@ -145,14 +140,14 @@ def get(self, exploration_id, thread_id):
                 'author_username': authors_settings[0].username,
                 'author_picture_data_url': (
-                    authors_settings[0].profile_picture_data_url),
+                    authors_settings[0].profile_picture_data_url)
```
Just for future reference: the comma at the end is OK here.

arunabh98 replied:
Okay.

seanlip commented on core/domain/feedback_services_test.py:
```diff
@@ -401,9 +403,19 @@ def test_get_thread_summaries(self):
     def test_get_thread_summaries_load_test(self):
         # The speed of fetching the summaries of 100 threads having 5 messages
-        # should be less than 1 second.
+        # should be less than 1.5 second.
```
I am concerned that this number keeps creeping upwards. Why can't it be under 1 second? A factor of 5 over the expected time of 0.2 seconds seems more than sufficient padding...

arunabh98 replied:
Explained the reason in the docstring as mentioned in the other thread.

seanlip commented on core/templates/dev/head/domain/feedback_message/FeedbackMessageObjectFactory.js:
```diff
@@ -0,0 +1,58 @@
+ */
+
+oppia.factory('FeedbackMessageObjectFactory', [function() {
+  var FeedbackMessage = function(text, suggestionHtml, currentContentHtml,
```
When something enclosed within (...) spans multiple lines, break after the opening paren. Ditto below.

arunabh98 replied:
Noted and Done.

seanlip commented on core/templates/dev/head/domain/feedback_message/FeedbackMessageObjectFactory.js:
```diff
@@ -0,0 +1,58 @@
+
+oppia.factory('FeedbackMessageObjectFactory', [function() {
+  var FeedbackMessage = function(text, suggestionHtml, currentContentHtml,
+    description,authorUsername, authorPictureDataUrl, createdOn) {
```
There is a missing space after "description,".

arunabh98 replied:
Fixed.

seanlip commented on core/templates/dev/head/domain/feedback_message/FeedbackMessageObjectFactory.js:
```diff
@@ -0,0 +1,58 @@
+      new Date());
+  };
+
+  FeedbackMessage.createFromBackendDicts = function(
```
Should be: FeedbackMessage.createFromBackendDict. Handle the "list" part outside of this factory; it doesn't need to know that the thing it's converting is part of a list. The caller should take care of that.

arunabh98 replied:
Actually, I'm passing a list to this function as the argument - feedbackMessageBackendDicts.

seanlip replied:
I know. I'm saying, just pass one dict to this function. Let the caller take care of calling it once per dict and constructing the list.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/feedback_message/FeedbackMessageObjectFactorySpec.js:
```diff
@@ -0,0 +1,56 @@
+      'FeedbackMessageObjectFactory');
+  }));
+
+  it('should fetch the feedback message domain objects from backend ' +
```
What about tests for createNewMessage()?

arunabh98 replied:
Added.

seanlip commented on core/templates/dev/head/domain/feedback_thread/FeedbackThreadObjectFactory.js:
```diff
@@ -0,0 +1,84 @@
+    this.originalAuthorId = originalAuthorId;
+    this.lastUpdated = lastUpdated;
+    this.lastMessageText = lastMessageText;
+    this.totalNoOfMessages = totalNoOfMessages;
```
totalNumMessages or totalMessageCount

(Let's avoid "No" as an abbreviation for number. It means other things too.)

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/feedback_thread/FeedbackThreadObjectFactory.js:
```diff
@@ -0,0 +1,84 @@
+    this.threadId = threadId;
+  };
+
+  FeedbackThread.prototype.updateSummaryOnNewMessage = function(
```
From the point of view of a feedback thread, I do not understand this method name. Do you mean something like addNewMessage or addNewMessageToThread? That's what seems to be happening here, AFAICT.

If that's the intended meaning then I would suggest pulling out the last two assignments into their own method, something like markAllMessagesAsRead(). They're logically separate, and the thread object should not need to have context about what's going on in its clients.

arunabh98 replied:
This function was intended to update the summaries of the thread on the addition of a new message. Now the function name has been changed to addNewMessage. So now it reads threadSummary.addNewMessage which I think gives the intended meaning.

seanlip commented on core/templates/dev/head/domain/feedback_thread/FeedbackThreadObjectFactory.js:
```diff
@@ -0,0 +1,84 @@
+   domain objects.
+ */
+
+oppia.factory('FeedbackThreadObjectFactory', [function() {
```
Either we are using these objects in the exploration editor too, or we are not.

If we are, then should explorationId and explorationTitle really be stored here? They seem external to the thread. If the exploration editor changes the title, then it might be messy to update all the threads. Arguably the id and title should not be stored here at all.

If we aren't, then perhaps what this actually represents is a FeedbackThreadSummary object, and it should be renamed accordingly.

Similar comments apply to other frontend domain objects introduced in this PR.

arunabh98 replied:
Yes, I had meant it to be a summary object only. Changed the instances everywhere.

seanlip commented on core/templates/dev/head/domain/feedback_thread/FeedbackThreadObjectFactory.js:
```diff
@@ -0,0 +1,84 @@
+      explorationTitle, explorationId, threadId);
+  }
+
+  FeedbackThread.createFromBackendDicts = function(
```
Again, the feedback thread object should only deal with a single thread dict. Let the caller handle the "assembling into a list" part.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/feedback_thread/FeedbackThreadObjectFactorySpec.js:
```diff
@@ -0,0 +1,86 @@
+      'Sample exploration 1');
+    expect(feedbackThreads[1].explorationTitle).toEqual(
+      'Sample exploration 2');
+    // expect(feedbackThreads[0].originalAuthorId).toEqual(
```
?

arunabh98 replied:
Fixed.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -79,8 +79,8 @@ oppia.factory('UrlInterpolationService', [
-        // hyphens, underscores, or periods.
-        var VALID_URL_PARAMETER_VALUE_REGEX = /^(\w| |_|-|[.])+$/;
+        // hyphens, underscores, periods or the equal to symbol.
+        var VALID_URL_PARAMETER_VALUE_REGEX = /^(\w| |_|-|[.]|=)+$/;
```
Oh, no, I think this won't work. According to RFC 3986 the '=' character is reserved. You will need to escape it before using the UrlInterpolationService, and you can't modify this. Make sure to test end-to-end to ensure that everything still works, even with '=' symbols.

That said, I'm slightly confused. We are actually generating feedback thread IDs with bad chars in them? Looking at the code, I thought we were generating them only from a base64 charset, which should not include the '=' sign...

arunabh98 replied:
The thread id does contain two == sign in the middle which is why it is failing. I logged this value and checked.

Should we revert back the earlier method of not using URL interpolation and using concatenation?

seanlip replied:
That's weird. It sounds like a bug to me, actually. What is causing the `==` to be introduced in the first place (in the backend)?

arunabh98 replied:
generate_new_thread_id() doesn't look like it inserts `==` in the backend. Where else could it be coming from?

seanlip replied:
I don't know. That was my question... if we want to do this properly, we need to actually understand what's going on.

arunabh98 replied:
I found [this](https://stackoverflow.com/questions/6916805/why-does-a-base64-encoded-string-have-an-sign-at-the-end). `=` is required to pad the string out to the proper length

seanlip replied:
That's a good find! I think that is the cause.

OK, so it's not a bug and we have to deal with it. Reading the function it looks like there's some escaping going on at the bottom. So, if you add '=' to this list, can you verify that it gets escaped correctly and that everything works?

If so, I retract my earlier statement, and it's fine to add '=' to the list. Also maybe append to the comment "Note that these parameter values are escaped before being added to the URL" and add a Karma test to ensure that that's the case when '=' signs are used.

arunabh98 replied:
Hi @Sean, I understood the part about adding '=' to the list. But I didn't understand this phrase - `gets escaped correctly`. Could you please explain this a bit more?

seanlip replied:
Look up encodeURIComponent: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -16,84 +16,91 @@
-      FATAL_ERROR_CODES, HUMAN_READABLE_SUBSCRIPTION_SORT_BY_KEYS,
+      $scope, $rootScope, $window, $http, $modal,
+      EXPLORATIONS_SORT_BY_KEYS_AND_I18N_IDS,
+      SUBSCRIPTION_SORT_BY_KEYS_AND_I18N_IDS, FATAL_ERROR_CODES,
```
Nit: have the line breaks match those in the imports above. Easier to compare and make sure nothing's been left out by accident (otherwise minification would fail).

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -16,84 +16,91 @@
+    $scope.username = GLOBALS.username;
+    var threadIndex = null;
+
+    $scope.newMessage = {text: ''};
```
```
  $scope.newMessage = {
    text: ''
  };
```

arunabh98 replied:
done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -221,79 +238,83 @@ oppia.controller('LearnerDashboard', [
-    $scope.sortFeedbackFunction = function(entity) {
-      return entity[$scope.currentFeedbackThreadsSortType];
-    }
+    $scope.sortFeedbackThreadsFunction = function(feedbackThread) {
```
Still not sure what "function" this is referring to -- it's not returning a function. Do you mean something like: getValueOfFeedbackThreadSortKey()?

Same thing for the corresponding functions for other entities.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -209,7 +225,8 @@ oppia.controller('LearnerDashboard', [
     $scope.sortExplorationFunction = function(entity) {
```
entity? Be more specific if you can.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -221,79 +238,83 @@ oppia.controller('LearnerDashboard', [
+            $scope.threadSummaries[index].threadId === threadId) {
+          threadIndex = index;
+          var threadSummary = $scope.threadSummaries[index];
+          if (threadSummary.authorSecondLastMessage) {
```
All this stuff should be part of the domain object -- something like threadSummary.markAllMessagesAsRead(). (Also, since you've gone to the trouble of defining threadSummary, you might as well use it throughout.)

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -221,79 +238,83 @@ oppia.controller('LearnerDashboard', [
-          $scope.threadSummary.author_last_message);
-        $scope.threadSummary.author_last_message = $scope.username;
-        $scope.threadSummary.total_no_of_messages += 1
+        newMessageSummary = FeedbackMessageObjectFactory.createNewMessage(
```
Missing "var".

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -346,14 +367,15 @@ oppia.controller('LearnerDashboard', [
             $scope.remove = function() {
               /* eslint-disable max-len */
-              if (subSectionName === LEARNER_DASHBOARD_SUBSECTIONS.EXPLORATIONS) {
+              if (subSectionName ===
```
Here and elsewhere, 'subsection' is one word.

arunabh98 replied:
done!

seanlip commented:
Weird. Apparently my comments got submitted before I clicked submit...

Anyway, I had a general note: this is looking better, and the remaining comments should be fairly minor and easy to fix!

However, I do have one concern. The learner dashboard HTML and JS files are getting very long, and the JS file in particular seems to contain a lot of different things. I made a comment earlier about pulling out parts of them into directives, but this comment was not addressed. 

If further development continues on the dashboard, the dashboard needs to be broken down more into its component subparts for maintainability, and I recommend using directives for this. The best model to follow is the collection editor; don't follow the creator dashboard (which has a controller that is far too monolithic) or the exploration editor (which uses Jinja and sub-controllers a lot, two patterns we want to get away from). In general, controllers should not contain a lot of logic; they just act as a bridge between the UI and the logic part of the code.

That said, refactoring the learner dashboard completely may take some time. If you could pull some subparts out of the current learner dashboard controller (such as the feedback messages view) into a directive, I think that would probably be a sufficient improvement. It is fine to do this in a subsequent PR if you prefer.

arunabh98 committed:
Address review comments 3.

arunabh98 committed:
Fix minor errors.

arunabh98 commented:
@seanlip  I have fixed the rest of the review comments. One you raised about "Reply... " was that should it be in the language of the exploration? I would say that doesn't matter because it is ultimately a placeholder and since it is in the language that the user has selected (I have internationalized it) the learner will understand that he/she has to reply to the thread. So he/she will just start typing his/her reply and gives the reply in whatever language the user wants to.

Also, I understand your concerns about maintainability. I would like to refactor this in a separate PR. For now, I would like to focus on milestone 2.3 as I would like to complete it first in the next few days.

arunabh98 committed:
Fix function name.

arunabh98 commented:
Also @seanlip, could you look why the base_test.py is failing. I looked at the log results and I'm not able to figure out why it is failing. If you could help a bit, it would be great!

seanlip commented:
Re backend tests: it's something to do with i18n. I recommend that you run just the single test locally to verify you're looking at the correct thing, then read the test itself to see what it's testing.

seanlip commented on core/templates/dev/head/domain/feedback_thread/FeedbackThreadSummaryObjectFactory.js:
```diff
@@ -17,16 +17,16 @@
 
-oppia.factory('FeedbackThreadObjectFactory', [function() {
+oppia.factory('FeedbackThreadSummaryObjectFactory', [function() {
   var FeedbackThread = function(status, originalAuthorId, lastUpdated,
```
If this is now a FeedbackThreadSummary then this var (and all references to this factory) should be renamed accordingly.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/exploration_editor/feedback_tab/feedback_tab.html:
```diff
@@ -128,7 +128,12 @@
           </div>
 
-          <div>
+          <div style="color:#ff0000; margin-top: 7px;" ng-if="tmpMessage.status === 'ignored' || tmpMessage.status === 'not_actionable'">
```
You need to forbid submit as well if no reason is specified.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/feedback_thread/FeedbackThreadSummaryObjectFactory.js:
```diff
@@ -36,48 +36,49 @@ oppia.factory('FeedbackThreadObjectFactory', [function() {
+    this.lastMessageRead = true;
+  }
+
+  FeedbackThread.prototype.updateThreadSummaryOnAdditionOfNewMessage = function(
```
This ends up being called as threadSummary.updateThreadSummary... which is a little weird. Just say threadSummary.appendNewMessage() or similar.

arunabh98 replied:
Done!

seanlip reviewed (changes requested):
@Arunabh98 -- this is looking better, but please respond to my previous review comments. There are at least three questions there that have gone unanswered, so I don't feel like I can rely on statements such as "I have fixed the rest of the review comments" in the future.

In the future, I will wait to do reviews until all comments from the previous round have been addressed / replied to, so that things do not fall through the cracks. So please make sure to address each specific comment explicitly, or indicate somehow that it is on your radar (e.g. by including a list of checkboxes at the end of the issue with remaining things to do).

Re the language: OK, thanks for the explanation.

Re maintainability: OK, but please file an issue and self-assign it so that we do not lose track of it. It will need to be done before even more functionality is added to the frontend anyway.

arunabh98 committed:
Address review comments 4.

arunabh98 committed:
Fix lint errors.

seanlip commented on core/templates/dev/head/domain/feedback_message/FeedbackMessageSummaryObjectFactory.js:
```diff
@@ -30,24 +30,24 @@ oppia.factory('FeedbackMessageSummaryObjectFactory', [function() {
-      feedbackMessageBackendDict.author_picture_data_url,
-      feedbackMessageBackendDict.created_on);
+  FeedbackMessageSummary.createFromBackendDict = function(
+    FeedbackMessageSummaryBackendDict) {
```
The arg should start with a lowercase letter: feedbackMessageSummaryBackendDict.

seanlip commented on core/templates/dev/head/domain/feedback_thread/FeedbackThreadSummaryObjectFactory.js:
```diff
@@ -54,32 +54,32 @@ oppia.factory('FeedbackThreadSummaryObjectFactory', [function() {
-      feedbackThreadBackendDict.exploration_id,
-      feedbackThreadBackendDict.thread_id);
+  FeedbackThreadSummary.createFromBackendDict = function(
+    FeedbackThreadSummaryBackendDict) {
```
Ditto.

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -252,6 +252,8 @@ oppia.controller('LearnerDashboard', [
 
     $scope.onClickThread = function(
       threadStatus, explorationId, threadId, explorationTitle) {
+      console.log(threadId);
```
Please remove console logs.

seanlip commented on core/templates/dev/head/domain/feedback_message/FeedbackMessageSummaryObjectFactory.js:
```diff
@@ -0,0 +1,53 @@
+    this.createdOn = createdOn;
+  };
+
+  FeedbackMessageSummary.createNewMessage = function(newMessage, authorUsername,
```
I've mentioned this before, but when a (...) spans multiple lines please break after the opening '(' -- here and elsewhere.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/feedback_message/FeedbackMessageSummaryObjectFactory.js:
```diff
@@ -0,0 +1,53 @@
+  };
+
+  FeedbackMessageSummary.createFromBackendDict = function(
+    FeedbackMessageSummaryBackendDict) {
```
The arg should start with a lowercase letter: feedbackMessageSummaryBackendDict.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/feedback_message/FeedbackMessageSummaryObjectFactorySpec.js:
```diff
@@ -0,0 +1,58 @@
+  }));
+
+  it('should create a new message', function() {
+    feedbackMessageSummary = (
```
Missing var

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/feedback_message/FeedbackMessageSummaryObjectFactorySpec.js:
```diff
@@ -0,0 +1,58 @@
+
+  it('should fetch the feedback message domain object from the backend ' +
+     'summary dict', function() {
+    messageSummary = {
```
Ditto

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/feedback_message/FeedbackMessageSummaryObjectFactorySpec.js:
```diff
@@ -0,0 +1,58 @@
+      created_on: 1000
+    };
+
+    feedbackMessageSummary = (
```
Ditto

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/feedback_thread/FeedbackThreadSummaryObjectFactory.js:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  FeedbackThreadSummary.createFromBackendDict = function(
+    FeedbackThreadSummaryBackendDict) {
```
Ditto: start with lowercase letter.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/feedback_thread/FeedbackThreadSummaryObjectFactory.js:
```diff
@@ -0,0 +1,85 @@
+    status, originalAuthorId, lastUpdated, lastMessageText, totalMessageCount,
+    lastMessageRead, secondLastMessageRead, authorLastMessage,
+    authorSecondLastMessage, explorationTitle, explorationId, threadId) {
+    return new FeedbackThreadSummary(status, originalAuthorId, lastUpdated,
```
Ditto, break after opening paren.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -170,60 +215,173 @@ oppia.controller('LearnerDashboard', [
+
+    $scope.onClickThread = function(
+      threadStatus, explorationId, threadId, explorationTitle) {
+      console.log(threadId);
```
Please remove console.logs

arunabh98 replied:
Done!

seanlip reviewed (changes requested):
Figured I'd put all the remaining stuff in one review so it's easy to reference -- sorry about the duplicate comments. I deleted the older ones. Note that the UrlInterpolationService thing also needs resolution.

Btw, this is looking great! If your mentors are happy with the UI/UX and the above comments are addressed then I think this is ready for merge.

arunabh98 committed:
Address review comments 5.

arunabh98 commented:
I have not yet addressed all review comments, specifically - Url interpolation service and the backend tests. I'll inform when I'm done with those.

jaredsilver reviewed (changes requested):
Thanks, @Arunabh98!

Just a couple other points:

1. Can we use `resize: vertical` instead of `resize: none` on the textbox so users can still resize it vertically?

2. It seems that the status updates are behaving a bit odd on the learner side. Specifically, if a creator updates the status to one thing, and then updates it to a second thing, all of the status messages on the learner side then end up saying the second thing (instead of the first one saying the first thing and the second one saying the second thing):

<img width="581" alt="screen shot 2017-07-24 at 1 17 52 pm" src="https://user-images.githubusercontent.com/1312199/28535588-eb3a46d6-7072-11e7-826a-b6e472afd7fc.png">

3. It seems there were a couple style-related regressions (perhaps related to the margin change, in which case we can just revert that and revisit it another time because there might be some broader style changes we'd want to make).

<img width="440" alt="screen shot 2017-07-24 at 1 17 10 pm" src="https://user-images.githubusercontent.com/1312199/28535624-0ed848a4-7073-11e7-884f-a5475c7493a1.png">
<img width="289" alt="screen shot 2017-07-24 at 1 17 19 pm" src="https://user-images.githubusercontent.com/1312199/28535623-0ed46d10-7073-11e7-83ff-1e956e38f994.png">

That's all I have. Great work!

arunabh98 committed:
Partially address review comments.

arunabh98 committed:
Fix status updates.

arunabh98 committed:
Fix lint errors.

seanlip commented on core/templates/dev/head/pages/learner_dashboard/learner_dashboard.html:
```diff
@@ -529,11 +532,17 @@ <h1 class="stat-value"><[subscriptionsList.length]></h1>
+                          <em translate="I18N_LEARNER_DASHBOARD_FEEDBACK_THREAD_STATUS_CHANGE_MESSAGE"
+                              translate-values="{threadStatus: getHumanReadableStatus(message.updatedStatus)}"></em>
+                        </span>
+                        <div ng-if="message.messageId !== 0 && message.updatedStatus">
```
Is anything supposed to go in this div?

arunabh98 replied:
I'm not yet done with this PR. Could you please wait till I send a message telling it's ready for another review?

seanlip commented:
@Arunabh98 -- this needs to go in pretty soon; if anything, the backend tests keep breaking on Travis for all the other PRs. What's left?

arunabh98 commented:
The URL interpolation service comment is left and the one base_test.py test keeps failing. It tells a key "I18N_IDS" in learner_dashboard.html is missing. I am not able to find the missing key in learner_dashboard.html

arunabh98 commented:
I'll finish this PR in the next hour or so.

seanlip commented:
Take a look at extract_keys_from_html_file() in the test. You may need to modify the regex pattern to start with a word break so that it only catch I18N keys and not variables/constants which contain _I18N_IDS in them..

arunabh98 committed:
Fix backend test, make threads responsive.

arunabh98 committed:
fix merge conflicts.

jaredsilver reviewed (approved):
LGTM from a UI/UX perspective! 👍

seanlip commented on core/controllers/base_test.py:
```diff
@@ -329,7 +329,7 @@ def _extract_keys_from_json_file(self, filename):
 
     def _extract_keys_from_html_file(self, filename):
-        regex_pattern = r'(I18N_[A-Z/_\d]*)'
+        regex_pattern = r'(I18N_[A-Z/_\d]*)(?=.*_I18N_IDS$).*$'
```
I think this no longer captures the stuff it used to capture.

Why not just add a `\b` to the front of the original regex? Would that work?

seanlip replied:
Maybe worth adding a comment to explain what this is trying to do, too, so that future devs don't break it.

arunabh98 replied:
Oops. Forgot to add the comment. Adding it...

arunabh98 replied:
Done!

arunabh98 committed:
Fix merge conflicts and fix regex pattern.

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -293,9 +457,19 @@ oppia.controller('LearnerDashboard', [
-        $scope.activeSection = LEARNER_DASHBOARD_SECTIONS.INCOMPLETE;
-        $scope.activeSubSection = LEARNER_DASHBOARD_SUBSECTIONS.EXPLORATIONS;
+        $scope.activeSection = LEARNER_DASHBOARD_SECTION_I18N_IDS.INCOMPLETE;
+        $scope.activeSubSection = (
```
Remember, here and elsewhere: "subsection" is one word, so this should be activeSubsection. There are other instances of this.

arunabh98 replied:
Done!

seanlip reviewed (approved):
LGTM! Did you verify the url interpolation thing? I had one more comment but otherwise feel free to merge when tests pass.

Thanks!

arunabh98 committed:
Add empty thread message, fix variable name, add comment.

arunabh98 committed:
Check = sign gets escaped correctly.

arunabh98 committed:
Fix abckend test.

seanlip commented:
OK, looks good! Is this mergeable? If so, please feel free to do so.

Btw, if that backend load test is causing problems, let's pull it out into a separate load_tests.py file that can be invoked somehow apart from the standard test system, or add an arg to the test script specifying whether to also run load tests and a flag to the test specifying whether it's a load test, or something. It's good to have the load test but we can't have this clogging up Travis, and I suspect something weird is going on because the test runtimes I've seen are always just above 1 sec (so somehow the test runtime depends on the limit, which makes no sense to me).

seanlip commented:
Oh, I forgot to say -- let's do the backend test pullout thing in a separate PR. It would be great if you could file an issue for that and self-assign it; we should get this in.

arunabh98 commented:
Okay I'll do this in a separate PR.

arunabh98 commented:
Even I am a bit confused with the results. I'll analyse what is happening in the PR for pulling the back end test out.

seanlip commented:
Awesome. Thanks @Arunabh98! And, with that, I think you've completed the milestone for month 2 -- congrats :D

Also, don't forget to file an issue for extracting the load tests -- we need to keep track of that.

Thanks!
